### PR TITLE
Feature/impact funcs

### PR DESCRIPTION
**Pull request by samluethi**
As discussed -  as a left-over from the wildfire branch, I've created two very simple default impact functions (a step function and a sigmoid function) in climada.entity.impact_funcs.base

### CODE REVIEW

File: climada/entity/impact_funcs/base.py [hunk 1]
```diff
@@ -127,4 +127,65 @@ def check(self):
...
+                maximum value of intensity range
+            if_id : int, optional, default=1
+                impact function id
+
+        """
 
+        self.id = if_id
+        self.intensity = np.array([inten_min, threshold, threshold, inten_max])
+        self.paa = np.array([1, 1, 1, 1])
+        self.mdd = np.array([0, 0, 1, 1])
```
Comment (chahank) [inline climada/entity/impact_funcs/base.py:154]: I suggest to add the possibility to choose the min and max impact too. 
```suggestion
    def set_step_ImpF(self, intensity, mdd=(0, 1), paa=(1, 1), impf_id=1):

        """ Step function type impact function. 
        
        By default, everything is destroyed above the step.
        Useful for high resolution modelling.
        
        This method modifies self (climada.entity.impact_funcs instance)
        by assigning an id, intensity, mdd and paa to the impact function.
        
        Parameters
        ----------
        intensity: tuple(float, float, float)
            tuple of 3-intensity numbers: (minimum, threshold, maximum)
        mdd: tuple(float, float)
            (min, max) mdd values. The default is (0, 1)
        paa: tuple(float, float)
            (min, max) paa values. The default is (1, 1)
        impf_id : int, optional, default=1
            impact function id

        """

        self.id = impf_id
        inten_min, threshold, inten_max = intensity
        self.intensity = np.array([inten_min, threshold, threshold, inten_max])
        paa_min, paa_max = paa
        self.paa = np.array([paa_min, paa_min, paa_max, paa_max])
        mdd_min, mdd_max = mdd
        self.mdd = np.array([mdd_min, mdd_min, mdd_max, mdd_max])
```

Comment (chahank) [reply climada/entity/impact_funcs/base.py:154]: The tabbing is a bit off. So, if you like the suggestions can you please clean it up?

File: climada/entity/impact_funcs/base.py [hunk 2]
```diff
@@ -127,4 +127,65 @@ def check(self):
...
+
+    def set_sigmoid_ImpF(self, sig_mid, sig_shape, sig_max,
```
Comment (chahank) [inline climada/entity/impact_funcs/base.py:156]: I do not understand the name `sig_shape`. Why not simply use `k`, the exponent of the sigmoid?

File: climada/entity/impact_funcs/base.py [hunk 3]
```diff
@@ -127,4 +127,65 @@ def check(self):
...
+                    inten_min, inten_max, inten_step=5, if_id=1):
+
+        """ Sigmoid type impact function hinging on three parameter. This type
+        of impact function is very flexible for any sort of study/resolution.
+        Parameters can be thought of as intercept (sig_mid), slope (sig_shape)
+        and top (sig_max) of a sigmoid. More precisely, sig_mid refers to the
+        intensity value where MDD equals 50% of sig_max.
```
Comment (chahank) [inline climada/entity/impact_funcs/base.py:163]:
```suggestion
        """ Sigmoid type impact function hinging on three parameter. 
        
        This type of impact function is very flexible for any sort of study/resolution.
        Parameters can be thought of as intercept (sig_mid), slope (sig_shape)
        and top (sig_max) of a sigmoid. More precisely, sig_mid refers to the
        intensity value where MDD equals 50% of sig_max.
```

File: climada/entity/impact_funcs/test/test_base.py [hunk 4]
```diff
@@ -35,6 +35,23 @@ def test_calc_mdr_pass(self):
...
         new_inten = 17.2
         self.assertEqual(imp_fun.calc_mdr(new_inten), 0.029583999999999996)
+        
+    def test_set_step(self):
+        """Check default impact function: step function"""
+        imp_fun = ImpactFunc()
+        imp_fun.set_step_ImpF(5, 0, 10)
+        self.assertTrue(np.array_equal(imp_fun.paa, np.ones(4)))
+        self.assertTrue(np.array_equal(imp_fun.mdd, np.array([0, 0, 1, 1])))
+        self.assertTrue(np.array_equal(imp_fun.intensity, np.array([0, 5, 5, 10])))
```
Comment (chahank) [inline climada/entity/impact_funcs/test/test_base.py:45]: If the generalization proposal is accepted this should be updated.

**Comment by chahank** [comment]
Great! Thanks for the addition of these template impact functions.

File: climada/entity/impact_funcs/base.py [hunk 5]
```diff
@@ -127,4 +127,65 @@ def check(self):
                            'origin of the intensity scale. In impact.calc '
                            'the impact is always null at intensity = 0.')
 
+    def set_step_ImpF(self, threshold, inten_min, inten_max, if_id=1):
```
Comment (Evelyn-M) [inline climada/entity/impact_funcs/base.py:130]: the naming is not consistent with the lynting conventions. change to 
```suggestion
    def set_step_impf(self, threshold, inten_min, inten_max, if_id=1):
```
 ?

File: climada/entity/impact_funcs/base.py [hunk 6]
```diff
@@ -127,4 +127,65 @@ def check(self):
...
+
+        """ Step function type impact function. Everything is destroyed above
+        threshold. Usefull for high resolution modelling.
```
Comment (Evelyn-M) [inline climada/entity/impact_funcs/base.py:133]:
```suggestion
        threshold. Useful for high resolution modelling.
```

Comment (Evelyn-M) [reply climada/entity/impact_funcs/base.py:133]: Why is a step function useful for high resolution modeling?

File: climada/entity/impact_funcs/base.py [hunk 7]
```diff
@@ -127,4 +127,65 @@ def check(self):
...
+
+        """
 
+        self.id = if_id
+        self.intensity = np.array([inten_min, threshold, threshold, inten_max])
+        self.paa = np.array([1, 1, 1, 1])
+        self.mdd = np.array([0, 0, 1, 1])
+
+    def set_sigmoid_ImpF(self, sig_mid, sig_shape, sig_max,
+                    inten_min, inten_max, inten_step=5, if_id=1):
```
Comment (Evelyn-M) [inline climada/entity/impact_funcs/base.py:157]: if the first function is modified that way, I'd also replace the `inten_min`, `inten_max`, `inten_step` with 
`intensity: tuple(float, float, float)
	            tuple of 3-intensity numbers: (minimum, threshold, maximum)`

File: climada/entity/impact_funcs/base.py (hunk 2 above)
Comment (Evelyn-M) [reply climada/entity/impact_funcs/base.py:156]: I also do not understand the sig_mid, sig_shape, sig_max. Usually, like in the TC sigmoids, we label the defining params sth like  v_min, v_half, v_max, (not exactly sure what now, but more or less)? so maybe call them thresh_...?
Especially the sig_mid as "intercept" i don't understand from the description.

Comment (samluethi) [reply climada/entity/impact_funcs/base.py:161]: as discussed - I don't want v_* as parameters as we want to keep this impact functions as generic as possible and v has always a windspeed touch to it. Also, this sigmoid is not the Emanuel function thus I don't want to use "his" labeling of params.
But obviously it's confusing as it is now so I try to improve ;)

File: climada/entity/impact_funcs/base.py (hunk 5 above)
Comment (samluethi) [reply climada/entity/impact_funcs/base.py]: why not. I don't think we're very consistent with this throughout climada and quite many use ImpF as abbreviation. but surely nice to be in line with lynting conventions. See 67385c69

File: climada/entity/impact_funcs/base.py (hunk 6 above)
Comment (samluethi) [reply climada/entity/impact_funcs/base.py]: as high resolution modelling means "hit-or-miss" for many hazards (i.e. wildfire, landslides, flooding,...). for that step functions are quite useful (i.e. in our calibration for wildfire, our sigmoid function converged to a step on 1-km resolution)

File: climada/entity/impact_funcs/base.py (hunk 7 above)
Comment (samluethi) [reply climada/entity/impact_funcs/base.py:162]: I'm adjusting to the tuple, but I do (minimum, maximum, step) to be in line with the np.arange naming of things

File: climada/entity/impact_funcs/base.py (hunk 2 above)
Comment (samluethi) [reply climada/entity/impact_funcs/base.py]: ok I renamed the params to L, k, and x0 to be in line with the wiki definition of the sigmoid. also, I've included the function in the docstring for "easier" understanding:
.. math::
            f(x) = \frac{L}{1+exp^{-k(x-x0)}}
.. and adjusted accordingly in 944f9aa7

File: climada/entity/impact_funcs/base.py (hunk 7 above)
Comment (samluethi) [reply climada/entity/impact_funcs/base.py]: see 944f9aa7

File: climada/entity/impact_funcs/test/test_base.py (hunk 4 above)
Comment (samluethi) [reply climada/entity/impact_funcs/test/test_base.py]: done in ef360bf3

**Comment by samluethi** [comment]
thanks a lot for the feedback @chahank and @Evelyn-M!

**Pull request merged**